### PR TITLE
[varLib.instancer] implement restricting axis ranges (aka L3) 

### PR DESCRIPTION
**Pull request by anthrotype**
This PR adds support for Level-3 partial instacing (as defined in https://github.com/fonttools/fonttools/issues/1537#issuecomment-474983825): i.e. restricting axes min/max coordinates, dropping delta regions that fall completely outside the new limits.

The core method is `limitTupleVariationAxisRange`. This is called for each `{axisTag: (min, max)}` range and for each TupleVariation table, and either drops it entirely if outside the range, or scales the deltas and updates the tent's min/default/max.

The command line tool `fonttools varLib.instancer` now accepts (colon-separated) axis ranges, in addition to single points. E.g. to limit the wght axis between 400:700 and drop-and-pin the wdth axis to 75, you do:

```
$ fonttools varLib.instancer wght=400:700 wdth=75
```

### CODE REVIEW

**Comment by anthrotype** [comment]
Note that changing the default is not currently supported; only changing the min and/or max is.
You currently get an error if the current default value is outside the requested range.

**Comment by davelab6** [comment]
Maybe the range should be specified with the same syntax as https://developers.google.com/fonts/docs/css2  so 400.. 700 instead of 400:700

**Comment by behdad** [comment]
Or a dash: 400-700.

I came to review, stayed to bikeshed.

File: Lib/fontTools/varLib/instancer.py [hunk 1]
```diff
@@ -127,28 +203,109 @@ def instantiateTupleVariationStore(variations, location, origCoords=None, endPts
...
 
-    # drop TupleVariation if all axes have been pinned (var.axes.items() is empty);
-    # its deltas will be added to the default instance's coordinates
-    defaultVar = newVariations.pop(tuple(), None)
+def limitTupleVariationAxisRanges(variations, axisRanges):
+    for axisTag, axisRange in sorted(axisRanges.items()):
+        newVariations = []
+        for var in variations:
+            newVariations.extend(limitTupleVariationAxisRange(var, axisTag, axisRange))
+        variations = newVariations
```
[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/fontTools/varLib/instancer.py]: newVariations local variable is only defined if the for loop has run at least once.

File: Lib/fontTools/varLib/instancer.py [hunk 2]
```diff
@@ -527,10 +701,48 @@ def _featureVariationRecordIsUnique(rec, seen):
...
+        minValue > maxValue
+        or minValue > axisRange.maximum
+        or maxValue < axisRange.minimum
+    ):
+        # condition invalid or out of range
+        return
+
+    values = [minValue, maxValue]
+    for i, value in enumerate(values):
+        if value < 0:
```
[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/fontTools/varLib/instancer.py]: yes, they are similar, but I prefer to keep it like it is; I think it makes clearer what's doing.

File: Lib/fontTools/varLib/instancer.py [hunk 3]
```diff
@@ -127,28 +203,109 @@ def instantiateTupleVariationStore(variations, location, origCoords=None, endPts
...
+        axisRange = NormalizedAxisRange(*axisRange)
+
+    # skip when current axis is missing (i.e. doesn't participate), or when the
+    # 'tent' isn't fully on either the negative or positive side
+    lower, peak, upper = var.axes.get(axisTag, (-1, 0, 1))
+    if peak == 0 or lower > peak or peak > upper or (lower < 0 and upper > 0):
+        return [var]
+
+    negative = lower < 0
+    if negative:
```
[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/fontTools/varLib/instancer.py]: as before, I think it's clearer to spell out the two cases

**Pull request merged**